File: src/blocks/BasicBuildingBlock.php
<?php

namespace jmw\fabricad\blocks;

use jmw\fabricad\shapes\Shape;

/**
 * The abs
 * @author jmw
 *
 */
class BasicBuildingBlock
{
    
    /**
     * 
     * @var array
     */
    protected $config = array();
    
    /**
     * @var string
     */
    protected $name = "";
    
    /**
     * 
     * @var string
     */
    protected $description = "";
    
    /**
     * 
     * @var \jmw\fabricad\shapes\Shape
     */
    protected $shape = null;
    
    /**
     * 
     * @return \jmw\fabricad\shapes\Shape
     */
    public function render(): Shape
    {
        return $this->shape;
    }
    
    
    /**
     * 
     * @param string $name
     * @param array $config
     */
    public function __construct(string $name = '', $config = array())
    {
        $this->setConfig($config);
        $this->setName($name);
    }
    
    /**
     * Returns the name of the object
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }
    
    /**
     * Sets the name of the building block
     * 
     * @param string $name
     * @return BasicBuildingBlock
     */
    public function setName(string $name): BasicBuildingBlock
    {
        $this->name = $name;
        
        return $this;
    }
    
    /**
     * @return string $description
     */
    public function getDescription()
    {
        return $this->description;
    }

    /**
     * @param string $description
     * @return BasicBuildingBlock
     */
    public function setDescription(string $description): BasicBuildingBlock
    {
        $this->description = $description;
        
        return $this;
    }
    
    /**
     * @return array $config
     */
    public function getConfig(): array
    {
        return $this->config;
    }

    /**
     * @param array $config
     */
    public function setConfig($config = array()): BasicBuildingBlock
    {
        $this->config = $config;
        return $this;
    }
    
    /**
     * 
     * @return Shape
     */
    public function  getShape(): Shape
    {
        return $this->shape;
    }
    
    /**
     * 
     * @param Shape $shape
     * @return BasicBuildingBlock
     */
    public function setShape(Shape $shape): BasicBuildingBlock
    {
        $this->shape = $shape;
        return $this;
    }

 
    
}

File: src/blocks/BasicBuildingBlock.cc

#include "BasicBuildingBlock.h"

namespace fabricad::blocks {

  string BasicBuildingBlock::getId()
  {
    return id_;
  }

  BasicBuildingBlock* BasicBuildingBlock::setId(string id)
  {
    id_ = id;
    return this;
  }

  string BasicBuildingBlock::getType()
  {
    return type_;
  }


  BasicBuildingBlock* BasicBuildingBlock::setType(string type)
  {
    type_ = type;
    return this;
  }


  std::string BasicBuildingBlock::toString()
  {
    return toString("");
  }

  std::string BasicBuildingBlock::toString(std::string indent)
  {
    return
      indent + "Name: " + getName() + "\n" +
      indent + "Type: " + getType() + "\n" ;
  }

  polygon BasicBuildingBlock::getShape()
  {
    return shape_;
  }

  BasicBuildingBlock* BasicBuildingBlock::setShape(polygon const &shape)
  {
    shape_ = shape;
    return this;
  }

  shapelayer BasicBuildingBlock::getLayer(size_t layer)
  {
    if (layers_.empty()) {
      render();
    }

    return layers_.at(layer);
  }

  std::vector<shapelayer> BasicBuildingBlock::getLayers()
  {
    if (layers_.empty()) {
      render();
    }

    return layers_;
  }

  /**
   * Most simple render, just add our shape to the
   * first layer
   */
  void BasicBuildingBlock::render()
  {
    layers_.clear();
    shapelayer l;
    l.polygons.push_back(shape_);
    l.id = getId();
    l.name = getName();
    layers_.push_back(l);

    shapelayer b;
    b.id = getId() + "_bricks";
    b.name = getName() + ": Bricks";
    layers_.push_back(b);

    shapelayer c;
    c.id = getId() + "_cutouts";
    c.name = getName() + ": Cut outs";
    layers_.push_back(c);

    shapelayer d;
    d.id = getId() + "_other";
    d.name = getName() + ": Other elements";
    layers_.push_back(d);

  }

}
